src/Plutarch/Test/QuickCheck.hs

```diff
@@ -213,60 +326,120 @@ type family PLamArgs (p :: S -> Type) :: [Type] where
 class
     (PLamArgs p ~ args) =>
     HaskEquiv
+        (e :: Equality)
+        (par :: Partiality)
         (h :: Type)
         (p :: S -> Type)
         (args :: [Type])
     where
     haskEquiv :: h -> TestableTerm p -> NP Gen args -> Property
 
--- TODO: Shrinking support
-
--- | @since 2.0.0
+-- | @since 2.1.0
 instance
     forall
+        (e :: Equality)
+        (par :: Partiality)
         (ha :: Type)
         (hb :: Type)
         (pa :: S -> Type)
         (pb :: S -> Type)
         (hbArgs :: [Type]).
     ( PLamArgs pb ~ hbArgs
-    , HaskEquiv hb pb hbArgs
+    , HaskEquiv e par hb pb hbArgs
     , PLifted pa ~ ha
     , PLift pa
     , PShow pa
     ) =>
-    HaskEquiv (ha -> hb) (pa :--> pb) (TestableTerm pa ': hbArgs)
+    HaskEquiv e par (ha -> hb) (pa :--> pb) (TestableTerm pa ': hbArgs)
     where
     haskEquiv h (TestableTerm p) (g :* gs) =
-        forAll g $ \(TestableTerm x) -> haskEquiv (h $ plift x) (TestableTerm $ p # x) gs
+        forAll g $ \(TestableTerm x) -> haskEquiv @e @par (h $ plift x) (TestableTerm $ p # x) gs
 
--- | @since 2.0.0
+instance
+    forall (e :: Equality) (p :: S -> Type) (h :: Type).
+    (HaskEquiv e 'ByComplete h p '[]) =>
+    HaskEquiv e 'ByPartial (Maybe h) p '[]
+    where
+    haskEquiv h (TestableTerm p) _ =
+        case evalTerm (Config{tracingMode = DoTracing}) p of
+            Left err ->
+                case h of
+                    Just _ -> failWith $ "Haskell expected success, but Plutarch compilation failed.\n" <> show err
+                    Nothing -> property True
+            Right (Left err, _, t) ->
+                case h of
+                    Just _ -> failWith $ "Haskell expected success, but Plutarch evaluation failed.\n" <> show err <> "\n" <> show t
+                    Nothing -> property True
+            Right (Right p', _, t) ->
+                case h of
+                    Just h' -> haskEquiv @e @( 'ByComplete) h' (TestableTerm p') Nil
+                    Nothing -> failWith $ "Haskell expected failure, but Plutarch succeed.\n" <> show t
```

Comment:
Do we consider a failure to _compile_ the same as a failure to execute with the expected answer? That seems a bit odd to me.

Reply:
I wasn't sure. Do you think I should make it fail regardless of expected partiality when compilation fails?

Reply:
This is a bit debatable. On the one hand, I would argue _compilation_ failure is a rather different kind of problem to _execution_ failure. However, I'm not sure whether these things end up being equivalent in practice. Probably best to ask.

Reply:
I'll wait for Slack discussion for this.